esp32/ota: Implement ESP-IDF OTA functionality.

ekondayan opened the pull request:
UPDATE: The test completed successfully on NodeMCU ESP32 by ai-thinker with ESP-IDF v4.0, v4.1, v4.2, v4.3

Implemented new functions in esp32.Partition:
* mark_app_invalid_rollback_and_reboot()
* check_rollback_is_possible()
* app_description()
* app_state()
* ota_begin()
* ota_write()
* ota_write_with_offset() for ESP-IDF version >= 4.2
* ota_end()
* ota_abort() for ESP-IDF version >= 4.3

* create tests
* update documentation

For many commercial products, Over The Air updates are a very important and critical part. It must be reliable and should not brick the device. Writing a good OTA from scratch is a daunting task. For that reason the use of well tested and proven reliable libraries is much more preferable than the ones developed in the house.

USECASE
=======
I'm developing an industrial device where the OTA is an essential part of it. Since only a few functions from esp-idf are 
implemented (enough for hobby project but not enough for commercial project), I ended up duplicating the esp-idf ota functionality in python. 
The result was a module with a questionable quality. I tried to predict all the possible places where it could crash, but my gut was 
telling me that I could be missing something. So I decided to implement more of the the OTA functions from esp-idf and to use them in my OTA module.
I've rewritten my OTA module and replaced the redundant code with the implemented functions from esp-idf. This allowed me to reduced the size of the module significantly, increase the robustness of the code and on top of that now the code got much simpler and easier to maintain.
The total increase in size of the compiled app image is 2432 bytes. 

IMPLEMENTATION
==============
  Extend the esp32.Partition class where all the OTA related functions are 
prefixed with "ota_" and app related functions are  prefixed with "app_".

Example:
  from esp32 import Partition
  app_part = Partition(Partition.RUNNING)
  app_part.app_description()
  app_part.app_state()
  handle = app_part.ota_begin()
  app_part.ota_end(handle)

New functions:
  Partition.mark_app_invalid_rollback_and_reboot(cls)
  Partition.check_rollback_is_possible(cls)
  Partition.app_description(self)
  Partition.app_state(self)
  Partition.ota_begin(self, image_size = 0)
  Partition.ota_write(self, handle_in, data_in)
  Partition.ota_write_with_offset(self, handle_in, data_in, offset)
  Partition.ota_end(self, handle_in)
  Partition.ota_abort(self, handle_in) only for ESP-IDF version >= 4.3

BENEFITS
========
* no code duplication
* reduced code size
* cleaner and easier to maintain code
* esp-idf handles encrypted flash
* use a reliable and well tested library written by the creators of ESP32
* more robust OTA procedure
* better performance

CONS
====
* None :) (maybe the 2432 bytes overhead in the final compiled app image)

ekondayan commented:
Anybody? Am I the only one who find this useful?

beyonlo commented:
Hi! Nice work!

Is this OTA feature in the 1.17 version, or was committed in the mainstream to tests?

If not, are there a roadmap to have an OTA by on ESP32 MicroPython?

I would like to test it!

Thank you.

ekondayan commented:
Hi,
unfortunately it is still not implemented mainstream. We are the only two people here who find this useful. The PR was submitted 6 months ago and there was not even a single comment from the community for that time.  

:(

You can find an example in the test section: tests/esp32/partition_ota.py
You can find the documentation here: docs/library/esp32.rst
To test it, just copy the tests/esp32/partition_ota.py to your freshly flashed esp32 and run it.

kevinkk525 commented:
I'd find an OTA option very useful too and have been following this PR.

beyonlo commented:
@dpgeorge and/or @jimmo or someone else.

Please, could you do a bit attention for this important PR?

1. In my opinion OTA feature is a very useful feature. In real real life, in the pratical way, is an essential feature, because will be possible to update easily the MicroPython version and applications via WiFi on the ESP32, for example. 
2. Another great thing in this PR is this OTA implementation following the same and tested (I hope) OTA from ESP-IDF implementation - this is great for evolution of OTA on MicroPython, following always from ESP-IDF.

Could you please to check, review or suggest any changes on this PR? And if necessary, the @ekondayan can to do the necessary changes to this amazing feature enter in the mainstream, and maybe ready to MicroPython 1.18.

Thank you so much for your attention.

lovetheguitar commented:
We are also following this PR very closely. Would make lots of things a lot more maintainable here. 

Thanks for all your work @ekondayan! It is greatly appreciated. 😺

beyonlo commented:
@ekondayan Hello.

This PR will works just on ESP32, or will works in all Espressif chips that use the ESP-IDF, like as ESP32-S2, ESP32-S3, ESP32-C3 and ESP32-C6?

As I know, as this PR use the ESP-IDF OTA, so I think that this PR will works in all that chips, Am I right?

Did you tested it in other chip than ESP32?

Thank you

ekondayan commented:
Yes this is just a wrapper around the esp-idf. It should work on all chips supported by esp-idf.

beyonlo commented:
Hi @ekondayan 

> Yes this is just a wrapper around the esp-idf. It should work on all chips supported by esp-idf.

Just to clarify: this wrapper is around this ESP OTA? -> https://github.com/espressif/esp-idf/tree/master/examples/system/ota

If not, what this ESP OTA differ from another ESP OTA that you has wrapped?

Thank you in advance!

ekondayan commented:
I suppose that you refer to "ESP HTTPS OTA", which itself uses a lower level API.

This PR is about the lower level API, which can be found here -> [https://github.com/espressif/esp-idf/blob/ef98a36/components/app_update/include/esp_ota_ops.h](https://github.com/espressif/esp-idf/blob/ef98a36/components/app_update/include/esp_ota_ops.h)

Link to the documentation -> [https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/system/ota.html](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/system/ota.html)

> If not, what this ESP OTA differ from another ESP OTA that you has wrapped?

This is the only OTA I have wrapped so far :)

Commit:
esp32/ota: Implement ESP-IDF OTA functionality.

Add OTA update support to the Partition class, including:
 - ESP-IDF OTA API wrappers.
 - app metadata inspection
 - rollback management

Implemented new functions:
* mark_app_invalid_rollback_and_reboot()
* check_rollback_is_possible()
* app_description()
* app_state() returns integer, use APP_STATE_* constants
* ota_begin()
* ota_write()
* ota_write_with_offset() for ESP-IDF version >= 4.2
* ota_end()
* ota_abort() for ESP-IDF version >= 4.3

Added APP_STATE_* constants to Partition class:
* APP_STATE_NEW
* APP_STATE_PENDING_VERIFY
* APP_STATE_VALID
* APP_STATE_INVALID
* APP_STATE_ABORTED
* APP_STATE_UNDEFINED

* create tests
* update documentation

demetz commented:
Hi,

do you have a small sample project demonstrating the functionality?

ekondayan commented:
@demetz Have a look at the test file in this PR

beyonlo commented:
Hello @ekondayan  

> You can find an example in the test section: tests/esp32/partition_ota.py 
> To test it, just copy the tests/esp32/partition_ota.py to your freshly flashed esp32 and run it.

I did that but do not works, show some errors. I'm using ESP32 with OTA firmware https://micropython.org/download/esp32-ota/ 

Here the errors:

```
>>> import os
>>> os.uname()
(sysname='esp32', nodename='esp32', release='1.18.0', version='v1.18 on 2022-01-17', machine='4MB/OTA module with ESP32')
>>> import partition_ota
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "partition_ota.py", line 143, in <module>
AttributeError: 'Partition' object has no attribute 'app_state'
>>>
```

Well, just to a try, I commented that lines, and have other errors:

```
>>> import partition_ota
Partition copy: (0, 16, 65536, 1572864, 'ota_0', False) --> (0, 16, 65536, 1572864, 'ota_0', False)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "partition_ota.py", line 159, in <module>
  File "partition_ota.py", line 110, in copy_partition_incr
AttributeError: 'Partition' object has no attribute 'ota_begin'
>>> 
```

Am I doing something wrong?

Thank you.

ekondayan commented:
> Am I doing something wrong?
  Yes. 

The image you are using is not based on my pull request. To use the OTA functions that I've implemented in my pull request, you have to check out the branch `feature/ota` from my fork and then to compile the image yourself.

beyonlo commented:
Hello @ekondayan 

> The image you are using is not based on my pull request. To use the OTA functions that I've implemented in my pull request, you have to check out the branch `feature/ota` from my fork and then to compile the image yourself.

I used that branch but the results was `SKIP`. It looks like I'm not compiling your branch, but it says it's the correct branch (`feature/ota`) on the git below. Any idea what is the error?

```
pi@pi:~/esp/esp-idf$ git clone -b feature/ota --single-branch https://github.com/ekondayan/micropython.git
pi@pi:~/esp/esp-idf$ cd micropython/
pi@pi:~/esp/esp-idf/micropython$ git branch -a 
* feature/ota
  remotes/origin/feature/ota
pi@pi:~/esp/esp-idf/micropython$ make -C mpy-cross
...
pi@pi:~/esp/esp-idf/micropython$ cd ports/esp32
pi@pi:~/esp/esp-idf/micropython/ports/esp32$ make submodules
git submodule update --init ../../lib/berkeley-db-1.xx
Submodule 'lib/berkeley-db-1.xx' (https://github.com/pfalcon/berkeley-db-1.xx) registered for path '../../lib/berkeley-db-1.xx'
Cloning into '/home/pi/esp/esp-idf/micropython/lib/berkeley-db-1.xx'...
Submodule path '../../lib/berkeley-db-1.xx': checked out '35aaec4418ad78628a3b935885dd189d41ce779b'
pi@pi:~/esp/esp-idf/micropython/ports/esp32$ make
idf.py -D MICROPY_BOARD=GENERIC -B build-GENERIC  build
...

pi@pi:~/esp/esp-idf/micropython/ports/esp32$ /home/pi/.espressif/python_env/idf4.4_py3.8_env/bin/python ../../../components/esptool_py/esptool/esptool.py -p /dev/ttyUSB0 -b 460800 --before default_reset --after hard_reset --chip esp32  write_flash --flash_mode dio --flash_size detect --flash_freq 40m 0x1000 build-GENERIC/bootloader/bootloader.bin 0x8000 build-GENERIC/partition_table/partition-table.bin 0x10000 build-GENERIC/micropython.bin


>>> os.uname()
(sysname='esp32', nodename='esp32', release='1.18.0', version='v1.14-1291-gefc889cf8 on 2022-05-27', machine='ESP32 module with ESP32')
>>> import partition_ota
SKIP
>>> 
```

Thank you in advance!

beyonlo commented:
Any idea about that @ekondayan?

Thank you.

> Hello @ekondayan
> 
> > The image you are using is not based on my pull request. To use the OTA functions that I've implemented in my pull request, you have to check out the branch `feature/ota` from my fork and then to compile the image yourself.
> 
> I used that branch but the results was `SKIP`. It looks like I'm not compiling your branch, but it says it's the correct branch (`feature/ota`) on the git below. Any idea what is the error?
> 
> ```
> pi@pi:~/esp/esp-idf$ git clone -b feature/ota --single-branch https://github.com/ekondayan/micropython.git
> pi@pi:~/esp/esp-idf$ cd micropython/
> pi@pi:~/esp/esp-idf/micropython$ git branch -a 
> * feature/ota
>   remotes/origin/feature/ota
> pi@pi:~/esp/esp-idf/micropython$ make -C mpy-cross
> ...
> pi@pi:~/esp/esp-idf/micropython$ cd ports/esp32
> pi@pi:~/esp/esp-idf/micropython/ports/esp32$ make submodules
> git submodule update --init ../../lib/berkeley-db-1.xx
> Submodule 'lib/berkeley-db-1.xx' (https://github.com/pfalcon/berkeley-db-1.xx) registered for path '../../lib/berkeley-db-1.xx'
> Cloning into '/home/pi/esp/esp-idf/micropython/lib/berkeley-db-1.xx'...
> Submodule path '../../lib/berkeley-db-1.xx': checked out '35aaec4418ad78628a3b935885dd189d41ce779b'
> pi@pi:~/esp/esp-idf/micropython/ports/esp32$ make
> idf.py -D MICROPY_BOARD=GENERIC -B build-GENERIC  build
> ...
> 
> pi@pi:~/esp/esp-idf/micropython/ports/esp32$ /home/pi/.espressif/python_env/idf4.4_py3.8_env/bin/python ../../../components/esptool_py/esptool/esptool.py -p /dev/ttyUSB0 -b 460800 --before default_reset --after hard_reset --chip esp32  write_flash --flash_mode dio --flash_size detect --flash_freq 40m 0x1000 build-GENERIC/bootloader/bootloader.bin 0x8000 build-GENERIC/partition_table/partition-table.bin 0x10000 build-GENERIC/micropython.bin
> 
> 
> >>> os.uname()
> (sysname='esp32', nodename='esp32', release='1.18.0', version='v1.14-1291-gefc889cf8 on 2022-05-27', machine='ESP32 module with ESP32')
> >>> import partition_ota
> SKIP
> >>> 
> ```
> 
> Thank you in advance!

ekondayan commented:
Sorry for the delay.

It looks like your current partition name does not starts with `ota_`.
For the test to work, you have to have two partitions of subtype ota_1 and ota_2 named ota_1 and ota_2.
Take a look at `micropython/ports/esp32/partitions-16MiB-ota.csv`

More info:
[](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-guides/partition-tables.html)

ekondayan commented:
Another requirement is to enable app rollback in the sdkconfig file

`CONFIG_BOOTLOADER_APP_ROLLBACK_ENABLE=y`

w0ssh commented:
cvxzbnhgfd@gmail.com
لقد اخترقو حسابي و قامو بتغير كلشي

abraha2d commented:
@ekondayan Thanks for keeping this up to date, I'm using this successfully in this project. Hopefully it can get mainlined soon.

ekondayan commented:
@abraha2d Thank you for your positive feedback. I am glad that it is useful to you. 

@dpgeorge I've been keeping this PR up do date for almost 2 years. Is there a slight chance this PR to go mainline?

glenn20 commented:
> Another requirement is to enable app rollback in the sdkconfig file
> 
> `CONFIG_BOOTLOADER_APP_ROLLBACK_ENABLE=y`

This is now enabled in the v1.21.0 release. See #12475 (sorry - I missed this PR when I was preparing that PR).

Note - if you are interested in OTA, I also have a micropython OTA tool at <https://github.com/glenn20/micropython-esp32-ota> which uses the existing support for OTA in micropython. There is also a separate tool at <https://github.com/glenn20/mp-image-tool-esp32> which can (among other things) add OTA partition tables to micropython firmware images and flash storage on ESP32 devices.

projectgus commented:
This is an automated heads-up that we've just merged a Pull Request
that removes the STATIC macro from MicroPython's C API.

See https://github.com/micropython/micropython/pull/13763

A search suggests this PR might apply the STATIC macro to some C code. If it
does, then next time you rebase the PR (or merge from master) then you should
please replace all the `STATIC` keywords with `static`.

Although this is an automated message, feel free to @-reply to me directly if
you have any questions about this.

projectgus commented on docs/library/esp32.rst:
```diff
@@ -173,15 +173,154 @@ methods to enable over-the-air (OTA) updates.
+
+.. method:: Partition.app_state()
+
+    Returns the app state of a valid ota partition. It can be one of the following strings:
```
The more common pattern in MicroPython would be to return an integer from this function, and define these as constants on the Partition class - i.e. `Partition.APP_STATE_NEW`, `Partition.APP_STATE_VERIFY`, etc.

projectgus commented on docs/library/esp32.rst:
```diff
@@ -173,15 +173,154 @@ methods to enable over-the-air (OTA) updates.
+    If the "CONFIG_BOOTLOADER_APP_ROLLBACK_ENABLE" option is set, and a reset occurs without
+    calling either 
+    ``mark_app_valid_cancel_rollback()`` or ``mark_app_invalid_rollback_and_reboot()``
+    function then the application is rolled back.
```
Is there a difference between calling this API, versus simply restarting without calling `mark_app_valid_cancel_rollback()`? If we can get the same behaviour without adding a new function here then it might be better to leave it out?

projectgus reviewed (commented):
Thanks for keeping this PR up to date for so long, @ekondayan.

The state of OTA support in MicroPython has probably changed since you first opened the PR, but if I follow correctly MicroPython already has implemented OTA functionality for ESP32 and this PR adds three related areas of functionality:

1. New APIs for reading app information from the partition.
2. Additional rollback support for marking an update invalid or testing if rollback is possible.
3. A wrapper around the ESP-IDF OTA write API rather than writing to the partition directly.

Part 1 makes sense to me as useful functionality for managing OTA updates in a deployed application. You could probably put these into a separate PR and we could merge them pretty quickly.

Part 2 I have some inline questions about, but the `check_rollback_is_possible()` function definitely seems useful to avoid reboot loops.

Part 3, I'm not certain what the exact benefits of an `ota_*` API is here compared to manually writing the partition and then calling `part.set_boot()` to set it as the next boot partition.

Is the main difference that the OTA API will verify the image? If so, could we add this as an optional boolean `verify` argument on `set_boot()` and get the same functionality? If there are any other benefits, can you please explain them?

Finally, the code size report suggests this PR adds 1.6KB of static RAM usage (`+1632(data) +8(bss)]`). I'm not sure how or why, there's no static buffer that I can see, but that's a big impact for ESP32s without PSRAM so it'd be good if there was a way to avoid it.
